sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java

```diff
@@ -61,6 +63,14 @@ public long durationMs() {
   public abstract void init(Iterator<InternalRow> iters[]);
 
   /**
+   * Initializes from array of iterators of InternalRow.
+   */
+  public void init(int index, Iterator<InternalRow> iters[]) {
+    partitionIndex = index;
```

Comment:
Not sure about this. I just ran test like:

    test("sort merge join/sample") {
      val N = 2 << 20
      runBenchmark("sort merge join", N) {
        val df1 = sqlContext.range(N)
          .selectExpr(s"(id * 15485863) % ${N*10} as k1")
        val df2 = sqlContext.range(N)
          .selectExpr(s"(id * 15485867) % ${N*10} as k2")
        df1.join(df2, col("k1") === col("k2")).sample(true, 0.2).count()
      }

I didn't see failure happened.